src/server/tran_server.cpp

```diff
@@ -551,34 +551,35 @@ tran_server::connection_handler::disconnect_async (bool with_disc_msg)
 
   m_disconn_future = std::async (std::launch::async, [this, with_disc_msg]
   {
-    on_disconnecting (); // server-type specific jobs before disconnecting.
-    // m_conn is not nullptr since it's only set to nullptr here below once
-
     m_conn->stop_response_broker (); // wake up threads waiting for a response and tell them it won't be served.
     auto ulock_conn = std::unique_lock<std::shared_mutex> { m_conn_mtx };
     const std::string channel_id = get_channel_id ();
+
+    // This may call this function (disconnect_async) while digesting all requests.
+    // In that, the lock of m_state_mtx will be acquired again, leading to a deadlock. We have to clean incoming requests first.
+    m_conn->stop_incoming_communication_thread ();
+
+    ulock_conn.unlock ();
+
+    on_disconnecting (); // server-type specific jobs before disconnecting.
```

Comment:
Could you tell me why this function was moved here? If it wasn't moved, it seems like there would be no need to unlock the `ulock_conn`

Reply:
I don't see the necessity. It can be much simpler with the other change. Thanks.